sw/host/tests/crypto/kmac_kat/src/main.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use arrayvec::ArrayVec;
+use clap::Parser;
+use std::fs;
+use std::time::Duration;
+
+use serde::Deserialize;
+
+use cryptotest_commands::commands::CryptotestCommand;
+use cryptotest_commands::kmac_commands::{
+    CryptotestKmacCustomizationString, CryptotestKmacKey, CryptotestKmacMessage,
+    CryptotestKmacMode, CryptotestKmacRequiredTagLength, CryptotestKmacTag,
+};
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::execute_test;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::test_utils::rpc::{UartRecv, UartSend};
+use opentitanlib::uart::console::UartConsole;
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    // Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "10s")]
+    timeout: Duration,
+
+    #[arg(long, num_args = 1..)]
+    kmac_json: Vec<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct KmacTestCase {
+    vendor: String,
+    test_case_id: usize,
+    algorithm: String,
+    mode: usize,
+    key: Vec<u8>,
+    message: Vec<u8>,
+    customization_string: Vec<u8>,
+    tag: Vec<u8>,
+    result: bool,
+}
+
+const KMAC_CMD_MAX_MESSAGE_BYTES: usize = 256;
+
+// These values are limitations of OpenTitan's KMAC HWIP
+const KMAC_CMD_MAX_CUSTOMIZATION_STRING_BYTES: usize = 36;
+const KMAC_CMD_MAX_KEY_BYTES: usize = 64;
+
+fn run_kmac_testcase(
+    test_case: &KmacTestCase,
+    opts: &Opts,
+    transport: &TransportWrapper,
+    fail_counter: &mut u32,
+) -> Result<()> {
+    log::info!(
+        "vendor: {}, test case: {}",
+        test_case.vendor,
+        test_case.test_case_id
+    );
+    let uart = transport.uart("console")?;
```

Comment:
I'm not familiar enough with opentitanlib and Rust to say for sure, but would it be better to pass the `uart` obtained in `test_kmac` into the `run_kmac_testcase` function rather than getting it again on each test case? I'm not sure if this is going to have flow control enabled.

Reply:
Interesting. All the previous test harnesses have used this structure (I just copied the structure from the AES one). My guess is that it would just re-bind to the same UART, but it might be worth trying out your suggestion on one of the algorithms. If it turns out to be materially faster, I can put in a later PR to change it for all the algorithms at once.